### ISSUE TITLE
Temp ed25519 certs stoped working after 2.5.0

### ISSUE THREAD

**Issue by htol** [ronf/asyncssh#581]
I have centralized ssh server which adding temp key/cert to my ssh agent after authentication.  I triaged that last working version was 2.5.0. I suspect this commit  https://github.com/ronf/asyncssh/commit/2fb89a76a51f0ff421a35aefaa7e9603ac630c65 After this point cert not working anymore.

My list of keys looks like this after it.
```shell
ssh-add -l
256 SHA256:XXXX temporary key 1689535229 (ED25519)
256 SHA256:XXXX temporary key 1689535229 (ED25519-CERT)
```

In every test I use this simple snippet.
```python
import asyncio, asyncssh, sys, logging

logging.basicConfig(
    format="%(asctime)s %(levelname)s %(message)s",
    level=logging.DEBUG,
    stream=sys.stdout,
)

async def run_client():
    async with asyncssh.connect("serverip", port=22227) as conn:
        result = await conn.run('echo "Hello!"', check=True)
        print(result.stdout, end="")

asyncio.get_event_loop().run_until_complete(run_client())
```

Here is logs after upgrade to asyncssh==2.13.1 (non working)
```log
2023-07-16 20:19:18,912 INFO Opening SSH connection to serverip, port 22227
2023-07-16 20:19:18,995 INFO [conn=0] Connected to SSH server at serverip, port 22227
2023-07-16 20:19:18,995 INFO [conn=0]   Local address: clientip, port 48352
2023-07-16 20:19:18,995 INFO [conn=0]   Peer address: serverip, port 22227
2023-07-16 20:19:18,995 DEBUG [conn=0] Sending version SSH-2.0-AsyncSSH_2.13.1
2023-07-16 20:19:18,995 DEBUG [conn=0] Received version SSH-2.0-OpenSSH_8.7
2023-07-16 20:19:18,996 DEBUG [conn=0] Requesting key exchange
2023-07-16 20:19:18,996 DEBUG [conn=0] Received key exchange request
2023-07-16 20:19:18,998 DEBUG [conn=0] Beginning key exchange
2023-07-16 20:19:19,007 DEBUG [conn=0] Completed key exchange
2023-07-16 20:19:19,008 INFO [conn=0] Beginning auth for user myusername
2023-07-16 20:19:19,052 DEBUG [conn=0] Trying public key auth with ssh-ed25519 key
2023-07-16 20:19:19,060 DEBUG [conn=0] Trying public key auth with ssh-ed25519 key ### <<< not a cert for some reason and I think this is a problem
2023-07-16 20:19:19,067 INFO [conn=0] Auth failed for user myusername
2023-07-16 20:19:19,068 INFO [conn=0] Connection failure: Permission denied
2023-07-16 20:19:19,068 INFO [conn=0] Aborting connection
```

non working server side debug log
```
debug1: userauth-request for user myusername service ssh-connection method publickey [preauth]
debug1: attempt 1 failures 0 [preauth]
debug2: input_userauth_request: try method publickey [preauth]
debug2: userauth_pubkey: valid user myusername querying public key ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIFIrvSp5BrZkbvGbwtgon6hjuvcXfI4OZCirsnSEomay [preauth]
debug1: userauth_pubkey: test pkalg ssh-ed25519 pkblob ED25519 SHA256:yatuvDJC6DMYzl2efEeuMAfcKy5aGmdpPTlLH4m2+Yw [preauth]
debug3: mm_key_allowed: entering [preauth]
debug3: mm_request_send: entering, type 22 [preauth]
debug3: mm_key_allowed: waiting for MONITOR_ANS_KEYALLOWED [preauth]
debug3: mm_request_receive_expect: entering, type 23 [preauth]
debug3: mm_request_receive: entering [preauth]
debug3: mm_request_receive: entering
debug3: monitor_read: checking request 22
debug3: mm_answer_keyallowed: entering
debug1: KRL version 0 generated at 20181001T153253
debug2: ssh_krl_file_contains_key: checking KRL /etc/security/revoked_keys
debug3: mm_answer_keyallowed: publickey authentication test: ED25519 key is not allowed
Failed publickey for myusername from 10.252.192.208 port 48352 ssh2: ED25519 SHA256:yatuvDJC6DMYzl2efEeuMAfcKy5aGmdpPTlLH4m2+Yw
debug3: mm_request_send: entering, type 23
debug2: userauth_pubkey: authenticated 0 pkalg ssh-ed25519 [preauth]
debug3: user_specific_delay: user specific delay 0.000ms [preauth]
debug3: ensure_minimum_time_since: elapsed 0.295ms, delaying 5.796ms (requested 6.091ms) [preauth]
debug3: userauth_finish: failure partial=0 next methods="publickey" [preauth]
debug3: send packet: type 51 [preauth]
debug3: receive packet: type 2 [preauth]
debug3: Received SSH2_MSG_IGNORE [preauth]
debug3: receive packet: type 50 [preauth]
debug1: userauth-request for user myusername service ssh-connection method publickey [preauth]
debug1: attempt 2 failures 1 [preauth]
debug2: input_userauth_request: try method publickey [preauth]
debug2: userauth_pubkey: valid user myusername querying public key ssh-ed25519 AAAAIHNzaC1lZDI1NTE5LWNlcnQtdjAxQG9wZW5zc2guY29tAAAAIFJRQjFqASFUsU1YCYPkKr6iUxSsTz+q4tgcS9O4hUuBAAAAIFIrvSp5BrZkbvGbwtgon6hjuvcXfI4OZCirsnSEomayGD/EFaBXm5EAAAABAAAAC2F0c3lnYW5lbmtvAAACZwAAAAV1c2VycwAAAA5uZXR0b29sc0FkbWlucwAAABVwcml2YXRlY2xvdWRPcGVyYXRvcnMAAAAJcGRuc1VzZXJzAAAADHNlbnRyeUFkbWlucwAAAAtwdXBwZXRVc2VycwAAABJzZXJ2ZXJkYkluZnJhVXNlcnMAAAADY3ZzAAAAFmtpYmFuYV9mbG9nc2VhcmNoVXNlcnMAAAAXa2liYW5hX2xvZ3NfYm9va2luZ19jb20AAAAIa3ZtVXNlcnMAAAALaGFkb29wVXNlcnMAAAAVa2liYW5hX2RsYnNlYXJjaFVzZXJzAAAAGWtpYmFuYV9jcm9ubG9nc2VhcmNoVXNlcnMAAAAKY3JvbmFjY2VzcwAAABhraWJhbmFfbmVsbGllc2VhcmNoVXNlcnMAAAAPbW9uaXRvcmluZ1VzZXJzAAAAC3N0YXRpY1VzZXJzAAAAEWNvbnRyb2xwbGFuZVVzZXJzAAAAGWtpYmFuYV9saXZlbG9nc2VhcmNoVXNlcnMAAAAIYXBwVXNlcnMAAAAOc2VydmVyZGJHdWVzdHMAAAANY29udGFpbmVyc0RldgAAABpraWJhbmFfY3VsZGVzYWNzZWFyY2hVc2VycwAAAA1hbnNpYmxlX1VzZXJzAAAAD3NtYXJ0YXZEZXBsb3llcgAAABRraWJhbmFfY2xvdWRpYW5Vc2VycwAAAApldGFwcFVzZXJzAAAACmV2ZW50VXNlcnMAAAAIc3NoVXNlcnMAAAAWa
userauth_pubkey: type mismatch for decoded key (received 7, expected 3) [preauth]
debug2: userauth_pubkey: authenticated 0 pkalg ssh-ed25519 [preauth]
debug3: user_specific_delay: user specific delay 0.000ms [preauth]
debug3: ensure_minimum_time_since: elapsed 5.021ms, delaying 1.070ms (requested 6.091ms) [preauth]
debug3: userauth_finish: failure partial=0 next methods="publickey" [preauth]
debug3: send packet: type 51 [preauth]
Connection closed by authenticating user myusername clientip port 48352 [preauth]
debug1: do_cleanup [preauth]
debug3: PAM: sshpam_thread_cleanup entering [preauth]
debug1: monitor_read_log: child log fd closed
debug3: mm_request_receive: entering
debug1: do_cleanup
debug1: PAM: cleanup
debug3: PAM: sshpam_thread_cleanup entering
debug1: Killing privsep child 9960
```

Here is working logs on asyncssh==2.5.0
```
2023-07-16 20:39:15,301 INFO Opening SSH connection to serverip, port 22227
2023-07-16 20:39:15,387 INFO [conn=0] Connection to serverip, port 22227 succeeded
2023-07-16 20:39:15,388 INFO [conn=0]   Local address: clientip, port 58678
2023-07-16 20:39:15,388 DEBUG [conn=0] Requesting key exchange
2023-07-16 20:39:15,389 DEBUG [conn=0] Received key exchange request
2023-07-16 20:39:15,392 DEBUG [conn=0] Beginning key exchange
2023-07-16 20:39:15,401 DEBUG [conn=0] Completed key exchange
2023-07-16 20:39:15,402 INFO [conn=0] Beginning auth for user myusernmae
2023-07-16 20:39:18,565 DEBUG [conn=0] Trying public key auth with ssh-ed25519 key
2023-07-16 20:39:18,573 DEBUG [conn=0] Trying public key auth with ssh-ed25519-cert-v01@openssh.com key ### <<< cert
2023-07-16 20:39:18,593 DEBUG [conn=0] Signing request with ssh-ed25519-cert-v01@openssh.com key
2023-07-16 20:39:18,667 INFO [conn=0] Auth for user myusername succeeded
2023-07-16 20:39:18,668 DEBUG [conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
2023-07-16 20:39:18,668 INFO [conn=0, chan=0] Requesting new SSH session
2023-07-16 20:39:18,687 DEBUG [conn=0] Received unknown global request: hostkeys-00@openssh.com
2023-07-16 20:39:18,687 DEBUG [conn=0] Received debug message: cert: key options: agent-forwarding port-forwarding pty x11-forwarding
2023-07-16 20:39:18,728 DEBUG [conn=0] Received debug message: principals: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
2023-07-16 20:39:18,728 DEBUG [conn=0] Received debug message: cert: key options: agent-forwarding port-forwarding pty x11-forwarding
2023-07-16 20:39:18,729 DEBUG [conn=0] Received debug message: principals: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
2023-07-16 20:39:18,730 INFO [conn=0, chan=0]   Command: echo "Hello!"
2023-07-16 20:39:18,766 INFO [conn=0, chan=0] Received exit status 0
2023-07-16 20:39:18,767 INFO [conn=0, chan=0] Received channel close
2023-07-16 20:39:18,767 INFO [conn=0, chan=0] Channel closed
Hello!
2023-07-16 20:39:18,768 INFO [conn=0] Closing connection
2023-07-16 20:39:18,768 INFO [conn=0] Sending disconnect: Disconnected by application (11)
2023-07-16 20:39:18,769 INFO [conn=0] Connection closed
```

working server side debug log
```
debug1: userauth-request for user myusername service ssh-connection method publickey [preauth]
debug1: attempt 1 failures 0 [preauth]
debug2: input_userauth_request: try method publickey [preauth]
debug2: userauth_pubkey: valid user myusername querying public key ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIFIrvSp5BrZkbvGbwtgon6hjuvcXfI4OZCirsnSEomay [preauth]
debug1: userauth_pubkey: test pkalg ssh-ed25519 pkblob ED25519 SHA256:yatuvDJC6DMYzl2efEeuMAfcKy5aGmdpPTlLH4m2+Yw [preauth]
debug3: mm_key_allowed: entering [preauth]
debug3: mm_request_send: entering, type 22 [preauth]
debug3: mm_key_allowed: waiting for MONITOR_ANS_KEYALLOWED [preauth]
debug3: mm_request_receive_expect: entering, type 23 [preauth]
debug3: mm_request_receive: entering [preauth]
debug3: mm_request_receive: entering
debug3: monitor_read: checking request 22
debug3: mm_answer_keyallowed: entering
debug1: KRL version 0 generated at 20181001T153253
debug2: ssh_krl_file_contains_key: checking KRL /etc/security/revoked_keys
debug3: mm_answer_keyallowed: publickey authentication test: ED25519 key is not allowed
Failed publickey for myusername from clientip port 58678 ssh2: ED25519 SHA256:yatuvDJC6DMYzl2efEeuMAfcKy5aGmdpPTlLH4m2+Yw
debug3: mm_request_send: entering, type 23
debug2: userauth_pubkey: authenticated 0 pkalg ssh-ed25519 [preauth]
debug3: user_specific_delay: user specific delay 0.000ms [preauth]
debug3: ensure_minimum_time_since: elapsed 0.304ms, delaying 5.787ms (requested 6.091ms) [preauth]
debug3: userauth_finish: failure partial=0 next methods="publickey" [preauth]
debug3: send packet: type 51 [preauth]
debug3: receive packet: type 2 [preauth]
debug3: Received SSH2_MSG_IGNORE [preauth]
debug3: receive packet: type 50 [preauth]
debug1: userauth-request for user myusername service ssh-connection method publickey [preauth]
debug1: attempt 2 failures 1 [preauth]
debug2: input_userauth_request: try method publickey [preauth]
debug2: userauth_pubkey: valid user myusername querying public key ssh-ed25519-cert-v01@openssh.com AAAAIHNzaC1lZDI1NTE5LWNlcnQtdjAxQG9wZW5zc2guY29tAAAAIFJRQjFqASFUsU1YCYPkKr6iUxSsTz+q4tgcS9O4hUuBAAAAIFIrvSp5BrZkbvGbwtgon6hjuvcXfI4OZCirsnSEomayGD/EFaBXm5EAAAABAAAAC2F0c3lnYW5lbmtvAAACZwAAAAV1c2VycwAAAA5uZXR0b29sc0FkbWlucwAAABVwcml2YXRlY2xvdWRPcGVyYXRvcnMAAAAJcGRuc1VzZXJzAAAADHNlbnRyeUFkbWlucwAAAAtwdXBwZXRVc2VycwAAABJzZXJ2ZXJkYkluZnJhVXNlcnMAAAADY3ZzAAAAFmtpYmFuYV9mbG9nc2VhcmNoVXNlcnMAAAAXa2liYW5hX2xvZ3NfYm9va2luZ19jb20AAAAIa3ZtVXNlcnMAAAALaGFkb29wVXNlcnMAAAAVa2liYW5hX2RsYnNlYXJjaFVzZXJzAAAAGWtpYmFuYV9jcm9ubG9nc2VhcmNoVXNlcnMAAAAKY3JvbmFjY2VzcwAAABhraWJhbmFfbmVsbGllc2VhcmNoVXNlcnMAAAAPbW9uaXRvcmluZ1VzZXJzAAAAC3N0YXRpY1VzZXJzAAAAEWNvbnRyb2xwbGFuZVVzZXJzAAAAGWtpYmFuYV9saXZlbG9nc2VhcmNoVXNlcnMAAAAIYXBwVXNlcnMAAAAOc2VydmVyZGJHdWVzdHMAAAANY29udGFpbmVyc0RldgAAABpraWJhbmFfY3VsZGVzYWNzZWFyY2hVc2VycwAAAA1hbnNpYmxlX1VzZXJzAAAAD3NtYXJ0YXZEZXBsb3llcgAAABRraWJhbmFfY2xvdWRpYW5Vc2VycwAAAApldGFwcFVzZXJzAAAACmV2ZW50VXNlcnMA
debug1: userauth_pubkey: test pkalg ssh-ed25519-cert-v01@openssh.com pkblob ED25519-CERT SHA256:yatuvDJC6DMYzl2efEeuMAfcKy5aGmdpPTlLH4m2+Yw CA ED25519 SHA256:Qxm/6OVRdLCDPPrtw6tgKG4bNzCPLMK2i2V1HQJTfg8 [preauth]
debug3: mm_key_allowed: entering [preauth]
debug3: mm_request_send: entering, type 22 [preauth]
debug3: mm_key_allowed: waiting for MONITOR_ANS_KEYALLOWED [preauth]
debug3: mm_request_receive_expect: entering, type 23 [preauth]
debug3: mm_request_receive: entering [preauth]
debug3: mm_request_receive: entering
debug3: monitor_read: checking request 22
```

**Comment by ronf** [comment]
Thanks for the details. I've been able to reproduce this here, and I think I have a fix for it. You were right that commit 2fb89a7 was at fault here. The problem was specifically related to cases where a key with an associated certificate was imported into an SSH agent and later used by AsyncSSH. The resulting AsyncSSH key pair ended up with the wrong algorithm name, making it look like a plain public key instead of a certificate.

Try the following and see if this works for you:

```patch
diff --git a/asyncssh/agent.py b/asyncssh/agent.py
index a776283..83c5954 100644
--- a/asyncssh/agent.py
+++ b/asyncssh/agent.py
@@ -149,6 +149,18 @@ class SSHAgentKeyPair(SSHKeyPair):
         self._is_cert = is_cert
         self._flags = 0

+    @property
+    def has_cert(self):
+        """ Return if this key pair has an associated cert"""
+
+        return self._is_cert
+
+    @property
+    def has_x509_chain(self):
+        """ Return if this key pair has an associated X.509 cert chain"""
+
+        return False
+
     def set_certificate(self, cert: SSHCertificate) -> None:
         """Set certificate to use with this key"""

diff --git a/asyncssh/public_key.py b/asyncssh/public_key.py
index 10b8f3e..eac909e 100644
--- a/asyncssh/public_key.py
+++ b/asyncssh/public_key.py
@@ -2096,6 +2096,18 @@ class SSHKeyPair:

         return self._key_type

+    @property
+    def has_cert(self):
+        """ Return if this key pair has an associated cert"""
+
+        return bool(self._cert)
+
+    @property
+    def has_x509_chain(self):
+        """ Return if this key pair has an associated X.509 cert chain"""
+
+        return self._cert and self._cert.is_x509_chain
+
     def get_algorithm(self) -> str:
         """Return the algorithm associated with this key pair"""

@@ -2194,9 +2206,9 @@ class SSHKeyPair:

         self.sig_algorithm = sig_algorithm

-        if not self._cert:
+        if not self.has_cert:
             self.algorithm = sig_algorithm
-        elif self._cert.is_x509_chain:
+        elif self.has_x509_chain:
             self.algorithm = sig_algorithm

             cert = cast('SSHX509CertificateChain', self._cert)
```

**Comment by htol** [comment]
Thank you very much Ron. It's working.

**Comment by ronf** [comment]
Great - thanks for the confirmation, and for the help tracking down the issue! This is mow available in the "develop" branch as commit 5e6450a and will be included in the next release.

**Comment by oriordan** [comment]
Hi @ronf, as we're looking forward to have this patch - can you tell when the next release is due?
Thank you!

**Comment by ronf** [comment]
Sorry for the delay. I've been working on other projects recently. I'll see if I can get something out over this coming weekend.

**Comment by ronf** [comment]
This fix is now available in AsyncSSH 2.14.0.
